### PR TITLE
[opentitantool] Add non-SPI rescue protocol for bootstrapping

### PR DESCRIPTION
**Pull request by jesultra**
Add support for a serial protocol supported by Ti50 firmware for
flashing a replacement image.  In order to add this protocol a few
changes to generic code has been made:

1. Refactoring of bootstrap/mod.rs to remove assumption of SPI and
bootstrap pin being used for all protocols.

2. Minor modifications to io/uart.rs trait and implementations, to
ensure uniform behavior of read_timeout(), and to simplify callers use
of write().

Change-Id: I2a612abb2f2de0b2f1f72cdc7bf94c0a51ec99de

### CODE REVIEW

File: sw/host/opentitanlib/src/bootstrap/mod.rs [hunk 1]
```diff
@@ -77,36 +123,36 @@ impl Bootstrap {
                 unimplemented!();
             }
         };
-        Ok(Bootstrap {
-            protocol,
+        let myself = Bootstrap {
```
Comment (mcy) [inline sw/host/opentitanlib/src/bootstrap/mod.rs:126]: might be wise to give this a more descriptive name (like `bootstrap`) even if we are local to the `Bootstrap` type.

File: sw/host/opentitanlib/src/bootstrap/rescue.rs [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+
+        ensure!(
+            &payload[0..4] == &Self::MAGIC_HEADER,
+            RescueError::ImageFormatError
+        );
+
+        // Find second occurrence of magic value.
+        let min_addr = match payload[256..]
+            .chunks(256)
+            .position(|c| &c[0..4] == &Self::MAGIC_HEADER)
```
Comment (mcy) [inline sw/host/opentitanlib/src/bootstrap/rescue.rs:88]: has_prefix, perhaps?

File: sw/host/opentitanlib/src/bootstrap/rescue.rs [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+    /// Creates a new `Rescue` protocol updater from `options`.
+    pub fn new(_options: &BootstrapOptions) -> Self {
+        Self {}
+    }
+
+    fn read_char(&self, uart: &dyn Uart) -> Result<char> {
+        let mut buf = [0u8; 1];
+        match uart.read_timeout(&mut buf, Duration::from_millis(100)) {
+            Ok(1) => Ok(buf[0] as char),
+            Ok(_) => Err(anyhow!("")),
```
Comment (mcy) [inline sw/host/opentitanlib/src/bootstrap/rescue.rs:177]: Seems some error messages are missing...

File: sw/host/opentitanlib/src/bootstrap/rescue.rs [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+            _ => Err(anyhow!("")),
+        }
+    }
+
+    fn expect_string(&self, uart: &dyn Uart, s: &str) -> Result<()> {
+        let char_vec: Vec<char> = s.chars().collect();
+        let mut char_vec: &[char] = &char_vec;
+        while char_vec.len() > 0 {
```
Comment (mcy) [inline sw/host/opentitanlib/src/bootstrap/rescue.rs:185]: Shouldn't this be `for char in s.chars()`?

File: sw/host/opentitanlib/src/bootstrap/rescue.rs [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+                return Err(anyhow!(""));
+            }
+            char_vec = &char_vec[1..];
+        }
+        Ok(())
+    }
+
+    fn flush_rx(&self, uart: &dyn Uart) {
+        let mut response = [0u8; Frame::HASH_LEN];
+        loop {
```
Comment (mcy) [inline sw/host/opentitanlib/src/bootstrap/rescue.rs:196]: I think a more idiomatic way to write this might be...
```
while let Ok(len) = uart.read_timeout() {
  if len == 0 { break }
}
```

File: sw/host/opentitanlib/src/bootstrap/rescue.rs [hunk 6]
```diff
@@ -0,0 +1,352 @@
...
+    /// receiver could be "out of sync".  This is resolved by sending one byte at a time, and
+    /// observing when the chip sends a response (which will be a rejection due to checksum).
+    fn synchronize(&self, uart: &dyn Uart) -> Result<()> {
+        // Most likely, only a few "extra" bytes have been sent during initial negotiation.
+        // Send almost a complete block in one go, and then send each of the last 16 bytes one
+        // at a time, slowly enough to detect a response before sending the next byte.
+        uart.write(&[0u8; 1008])?;
+        let mut response = [0u8; 1];
+        let limit = match uart.read_timeout(&mut response, Duration::from_millis(50)) {
+            Ok(0) => 16,
```
Comment (mcy) [inline sw/host/opentitanlib/src/bootstrap/rescue.rs:215]: You can merge these arms using `Ok(0) | Err(_)`.

File: sw/host/opentitanlib/src/bootstrap/rescue.rs [hunk 7]
```diff
@@ -0,0 +1,352 @@
...
+        while char_vec.len() > 0 {
+            if self.read_char(uart)? != char_vec[0] {
+                return Err(anyhow!(""));
+            }
+            char_vec = &char_vec[1..];
+        }
+        Ok(())
+    }
+
+    fn flush_rx(&self, uart: &dyn Uart) {
```
Comment (mcy) [inline sw/host/opentitanlib/src/bootstrap/rescue.rs:194]: Is there a reason to not return an error?

File: sw/host/opentitanlib/src/bootstrap/rescue.rs [hunk 8]
```diff
@@ -0,0 +1,352 @@
...
+        // Attempt getting the attention of the bootloader.
+        let timeout = Duration::from_millis(2000);
+        for _ in 0..Self::MAX_CONSECUTIVE_ERRORS {
+            eprint!("Resetting...");
+            container.reset_pin.write(false)?; // Low active
+            std::thread::sleep(container.reset_delay);
+            container.reset_pin.write(true)?; // Release reset
+            let stopwatch = Instant::now();
+            while stopwatch.elapsed() < timeout {
+                if let Ok(_) = self.expect_string(&*uart, "Bldr |") {
```
Comment (mcy) [inline sw/host/opentitanlib/src/bootstrap/rescue.rs:251]: This condition may want to be
```
if self.expect_string(...).is_err() { continue }
```

File: sw/host/opentitanlib/src/bootstrap/rescue.rs [hunk 9]
```diff
@@ -0,0 +1,352 @@
...
+        transport: &TransportWrapper,
+        payload: &[u8],
+    ) -> Result<()> {
+        let frames = Frame::from_payload(payload)?;
+        let uart = container.uart_params.create(transport)?;
+
+        self.enter_rescue_mode(container, &*uart)?;
+
+        // Send frames one at a time.
+        'next_block: for transmit_index in 0..frames.len() {
```
Comment (mcy) [inline sw/host/opentitanlib/src/bootstrap/rescue.rs:305]: `for (idx, frame) in frames.iter().enumerate()`

File: sw/host/opentitanlib/src/bootstrap/rescue.rs [hunk 10]
```diff
@@ -0,0 +1,352 @@
...
+                        Ok(0) | Err(_) => break,
+                        Ok(n) => index += n,
+                    }
+                }
+                if index < Frame::HASH_LEN {
+                    eprint!("sync.");
+                    self.synchronize(&*uart)?;
+                    continue;
+                }
+                let response_chunks = response.chunks(4).collect::<Vec<_>>();
```
Comment (mcy) [inline sw/host/opentitanlib/src/bootstrap/rescue.rs:328]: I would have written this as
```rust
if response[4..].chunks(4).all(|&x| == response[..4]) { ... }
```

File: sw/host/opentitanlib/src/bootstrap/rescue.rs [hunk 11]
```diff
@@ -0,0 +1,352 @@
...
+                if response_chunks.iter().all(|&x| x == response_chunks[0]) {
+                    eprint!("sync.");
+                    self.synchronize(&*uart)?;
+                } else if response == frame.frame_hash() {
+                    continue 'next_block;
+                } else {
+                    self.flush_rx(&*uart);
+                    if consecutive_errors >= 3 {
```
Comment (mcy) [inline sw/host/opentitanlib/src/bootstrap/rescue.rs:336]: How is this constant related to MAX_CONSECUTIVE_ERRORS?

File: sw/host/opentitanlib/src/bootstrap/mod.rs [hunk 12]
```diff
@@ -33,42 +40,81 @@ arg_enum! {
...
 }
 
-impl Bootstrap {
+impl<'a> Bootstrap<'a> {
     const RESET_DELAY: Duration = Duration::from_millis(200);
 
-    /// Consrtuct a `Bootstrap` struct configured to use `protocol` and `options`.
-    pub fn new(protocol: BootstrapProtocol, options: BootstrapOptions) -> Result<Self> {
-        let updater: Box<dyn UpdateProtocol> = match protocol {
+    /// Perform the update, sending the firmware `payload` to the `spi` target,
```
Comment (timothytrippel) [inline sw/host/opentitanlib/src/bootstrap/mod.rs:107]: nit: Can this a UART target in the case of the new "rescue" protocol?

**Review by timothytrippel — APPROVED** [review]
One minor nit in addition to @mcy's comments. Otherwise, architecturally this LGTM (as we discussed offline).

File: sw/host/opentitanlib/src/bootstrap/mod.rs (hunk 1 above)
Comment (jesultra) [reply sw/host/opentitanlib/src/bootstrap/mod.rs:126]: I have factored this method, such that I can use the "self" reserved word.

File: sw/host/opentitanlib/src/bootstrap/rescue.rs (hunk 2 above)
Comment (jesultra) [reply sw/host/opentitanlib/src/bootstrap/rescue.rs:88]: As far as I can tell, strings have a starts_with() method, but I have not found an equivalent for array slices, are you proposing that I declare a helper method myself.  I feel that the expression `&c[0..4] == &Self::MAGIC_HEADER` is short enough that a helper method is hard to justify.  (Avoiding to specify the number 4 explicitly should be the only advantage.

File: sw/host/opentitanlib/src/bootstrap/rescue.rs (hunk 3 above)
Comment (jesultra) [reply sw/host/opentitanlib/src/bootstrap/rescue.rs:177]: I have rewritten these two methods to return Option<char> and bool, as the Err(...) returns were immediately handled by the caller, and was part of normal operation.

File: sw/host/opentitanlib/src/bootstrap/rescue.rs (hunk 4 above)
Comment (jesultra) [reply sw/host/opentitanlib/src/bootstrap/rescue.rs:185]: Yes, that is much nicer.

File: sw/host/opentitanlib/src/bootstrap/rescue.rs (hunk 5 above)
Comment (jesultra) [reply sw/host/opentitanlib/src/bootstrap/rescue.rs:196]: Your suggestion does not consider the possibility of Err(...) return from read_timeout, which I want to treat the same as Ok(0) as the likely error is timeout.  (Some implementation return Ok(0) on timeout, some return some Err(...).)

File: sw/host/opentitanlib/src/bootstrap/rescue.rs (hunk 6 above)
Comment (jesultra) [reply sw/host/opentitanlib/src/bootstrap/rescue.rs:215]: I have done so here, and above in flush_rx().

File: sw/host/opentitanlib/src/bootstrap/rescue.rs [hunk 13]
```diff
@@ -0,0 +1,352 @@
...
+                // A response at this point must mean that more than 16 bytes had already been
+                // sent before entering this method.  This will be resolved by doing another
+                // slower round of 1024 bytes with delay in between every one.
+                self.flush_rx(uart);
+                1024
+            }
+        };
+        for _ in 0..limit {
+            uart.write(&[0u8; 1])?;
+            match uart.read_timeout(&mut response, Duration::from_millis(50)) {
```
[the remark this answers is not in the document]
Comment (jesultra) [reply sw/host/opentitanlib/src/bootstrap/rescue.rs:227]: Longer term, I want to look at the various implementation of the Uart trait, and make sure that the read_timeout always treats an actual timeout as "normal" and returns Ok(0), reserving Err(_) for truly unexpected occurrences, such as device disconnect and the like.  With that I could use the ? operator and simple if conditions on the returned integer value.

I find it somewhat confusing that in your proposal the control flow on Err(_) takes a different path from Ok(0), I have instead combined the two into one match clause: Ok(0) | Err(_), then I hope to be able to do the refactoring that I mentioned above at a later point.

File: sw/host/opentitanlib/src/bootstrap/rescue.rs (hunk 7 above)
Comment (jesultra) [reply sw/host/opentitanlib/src/bootstrap/rescue.rs:194]: It is not clear to me whether this comment is meant for the flush_rx() or expect_string() method.

In the future, when "ordinary" timeouts do not use Err(...) in read_timeout, then we can bail out on any Err(_) from the underlying driver, and at that point, it would be useful to return Result<...> from both expect_string() and flush_rx().  Right now, it would be too much in my opinion, to try to distinguish the various error values.  Device disconnect or similar in the middle of an operation is pretty rare, and I would rather put the effort into the cleaning up Err() usage of the trait implementations.

File: sw/host/opentitanlib/src/bootstrap/rescue.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (jesultra) [reply sw/host/opentitanlib/src/bootstrap/rescue.rs:251]: Hm, it appears not.  (In some earlier iteration of the code, I did get errors without this.  I am not exactly sure what I changed.)

Comment (jesultra) [reply sw/host/opentitanlib/src/bootstrap/rescue.rs:251]: Done

File: sw/host/opentitanlib/src/bootstrap/rescue.rs (hunk 10 above)
Comment (jesultra) [reply sw/host/opentitanlib/src/bootstrap/rescue.rs:328]: If I try that, I get:
```
if response[4..].chunks(4).all(|&x| x == response[..4]) {
                                 ^ doesn't have a size known at compile-time
```

File: sw/host/opentitanlib/src/bootstrap/rescue.rs (hunk 11 above)
Comment (jesultra) [reply sw/host/opentitanlib/src/bootstrap/rescue.rs:336]: It is related in that some time before reaching the maximum number of retries, we want to take an additional measure, which it a little time consuming, but could resolve the problem.  I have added a named constant with some comment.

File: sw/host/opentitanlib/src/bootstrap/mod.rs (hunk 12 above)
Comment (jesultra) [reply sw/host/opentitanlib/src/bootstrap/mod.rs:107]: Oh yes, I forgot to update this comment, leaving it referring to parameters that no longer exist.  I have updated it.

File: sw/host/opentitanlib/src/bootstrap/rescue.rs (hunk 2 above)
Comment (mcy) [reply sw/host/opentitanlib/src/bootstrap/rescue.rs:88]: https://doc.rust-lang.org/std/primitive.slice.html#method.starts_with is what you want. :)

File: sw/host/opentitanlib/src/bootstrap/rescue.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (mcy) [reply sw/host/opentitanlib/src/bootstrap/rescue.rs:232]: Sounds good.

File: sw/host/opentitanlib/src/bootstrap/rescue.rs (hunk 7 above)
Comment (mcy) [reply sw/host/opentitanlib/src/bootstrap/rescue.rs:201]: Got it.

File: sw/host/opentitanlib/src/bootstrap/rescue.rs (hunk 10 above)
Comment (mcy) [reply sw/host/opentitanlib/src/bootstrap/rescue.rs:333]: Write
```rust
|x| x == &respose[..4]
```
I got that mixed up, sorry.

**Review by mcy — APPROVED** [review]
LG modulo a couple replies to existing threads.

File: sw/host/opentitanlib/src/bootstrap/rescue.rs (hunk 2 above)
Comment (jesultra) [reply sw/host/opentitanlib/src/bootstrap/rescue.rs:88]: Thank you, I do not know how I could miss that.

File: sw/host/opentitanlib/src/bootstrap/rescue.rs (hunk 10 above)
Comment (jesultra) [reply sw/host/opentitanlib/src/bootstrap/rescue.rs:333]: Done

**Comment by jesultra** [comment]
I have addressed all the remaining comments.  Tim, could you please merge this PR when convenient.

**Comment by timothytrippel** [comment]
Looks like this PR is still failing CI (seems like you didn't sign off the commit with a "git commit -s ..." to produce the right commit message). Could you please fix that @jesultra? Then I can merge.

Thanks!

**Comment by timothytrippel** [comment]
Hmmm, still failing the CI commit message check, see [here](https://github.com/lowRISC/opentitan/blob/master/util/lint_commits.py) for exactly what this check does. Additionally, you do not need the last line of the commit message "Change-Id: ..." (that is a Gerrit requirement, and not relevant to our VCS).

**Comment by jesultra** [comment]
Hm, I have added the "Signed-off-by", but I do not see what my change could have to do with "Verilog format" error, that I get from the CI.

**Comment by timothytrippel** [comment]
Oh it looks like this issue was due to some merge skew that happened within our CI earlier today. If you just rebase on master, and re-force push, it should pass CI.

Also, would you mind removing the "Change-Id:" line from the commit message? (this is not needed)

Sorry for the hassle on this, and thanks for the fixes!

**Review by timothytrippel — APPROVED** [review]
